src/System.Net.Utilities/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -2,33 +2,268 @@
+
+                    sw.Stop();
+                    long roundTripTime = sw.ElapsedMilliseconds;
+                    // We want to return a buffer with the actual data we sent out, not including the header data.
+                    byte[] dataBuffer = new byte[bytesReceived - IcmpHeaderLengthInBytes];
+                    Array.Copy(receiveBuffer, IcmpHeaderLengthInBytes, dataBuffer, 0, dataBuffer.Length);
+
+                    IPStatus status = isIpv4 
+                                        ? Icmpv4MessageConstants.MapV4TypeToIPStatus(type, code) 
+                                        : Icmpv6MessageConstants.MapV6TypeToIPStatus(type, code);
+                    
+                    return new PingReply(address, options, status, roundTripTime, dataBuffer);
+                }
+
+                // We have exceeded our tiemout duration, and no reply has been received.
+                sw.Stop();
+                throw new PingException(SR.net_ping_reply_timeout);
+            }
+        }
+
+        private PingReply SendWithPingUtility(IPAddress address, byte[] buffer, int timeout, PingOptions options)
+        {
+            bool isIpv4 = address.AddressFamily == AddressFamily.InterNetwork;
+            string pingExecutable = isIpv4 ? s_discoveredPing4UtilityPath : s_discoveredPing6UtilityPath;
+            if (pingExecutable == null)
+            {
+                throw new PlatformNotSupportedException(SR.net_ping_utility_not_found);
+            }
+
+            StringBuilder sb = new StringBuilder();
+            sb.Append("-c 1"); // Just send a single ping ("count = 1")
+
+            // The command-line flags for "Do-not-fragment" and "TTL" are not standard.
+            // In fact, they are different even between ping and ping6 on the same machine.
+
+            // The ping utility is not flexible enough to specify an exact payload.
+            // But we can at least send the right number of bytes.
+            sb.Append(" -s ");
+            sb.Append(buffer.Length);
+
+            sb.Append(" ");
+            sb.Append(address.ToString());
+
+            string processArgs = sb.ToString();
+            ProcessStartInfo psi = new ProcessStartInfo(pingExecutable, processArgs);
+            psi.RedirectStandardOutput = true;
+            psi.RedirectStandardError = true;
+            Process p = new Process() { StartInfo = psi };
+            p.Start();
+            if (!p.WaitForExit(timeout))
+            {
+                throw new PingException(SR.net_ping_reply_timeout);
+            }
+
+            if (p.ExitCode != 0)
+            {
+                // This means no reply was received, although transmission may have been successful.
+                throw new PingException(SR.net_ping_reply_timeout);
+            }
+
+            string output = p.StandardOutput.ReadToEnd();
+            int timeIndex = output.IndexOf("time=", StringComparison.Ordinal);
+            int afterTime = timeIndex + 5;
```

Comment:
I assume the 5 here is the length of "time="?  It'd be nice to put the string into a constant and then use Length off of it.